### PR TITLE
Implement updateOwnerStatus game command

### PR DESCRIPTION
**Pull request by GalBr**
This is my first PR in OpenLoco, and the first time I'm trying to reverse engineer a function here.
Any feedback and help making it better is appreciated.

Here are the things I noticed:
1. Do we need to check that the company returned from` CompanyManager` exists? It's not doing that in the original code, and it is happening in the calling function in `CompanyManager.cpp`.
2. I called the parameters to the function `param1` and `param2` because they can be either `position.x` and` position.y` or `-2` and `EntityId`. Is this OK?

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
You'll need to manually add the new file to the openloco.vcxproj file to get the windows ci to compile.

I was thinking about it and we could introduce a helper class for the two different uses of this and use that class in the company struct and the  game command:
```cpp
class OwnerStatus
{
    int16_t data[2];

public:
    OwnerStatus(EntityId id){ data[0] = -2; data[1] = enumValue(id);}
    OwnerStatus(){data[0] = -1; data[1] = 0;}
    OwnerStatus(const Map::Pos2& pos}{data[0] = pos.x; data[1] = pos.y;}

    constexpr bool isEntity() const { return data[0] == -2; }
    constexpr bool isPosition() const { return !isEntity() && !empty(); }
    [[nodiscard]] constexpr bool empty() const { return data[0] == -1; }

    Map::Pos2 getPosition(){ return Map::Pos2{ data[0], data[1] };
    EntityId getEntity(){ return static_cast<EntityId>(data[1]);
}
```

**Comment by GalBr** [comment]
I updated the branch with some new code. I'm not sure this is working correctly yet.

File: src/OpenLoco/GameCommands/GameCommands.h [hunk 1]
```diff
@@ -1577,25 +1578,33 @@ namespace OpenLoco::GameCommands
...
-        regs.ax = position.x;
-        regs.cx = position.y;
-        doCommand(GameCommand::updateOwnerStatus, regs);
-    }
+        OwnerStatus ownerStatus;
+
+        explicit operator registers() const
+        {
+            registers regs;
+            if (ownerStatus.isEntity())
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/GameCommands.h:1594]: you dont need to do this switch this is already handled by the constructor of the class. Just provide some sort of way to get the data.

File: src/OpenLoco/Company.h [hunk 2]
```diff
@@ -46,6 +47,41 @@ namespace OpenLoco
...
+            data[0] = -2;
+            data[1] = enumValue(id);
+        }
+        OwnerStatus(const Map::Pos2& pos)
+        {
+            data[0] = pos.x;
+            data[1] = pos.y;
+        }
+
+        constexpr bool isEntity() const { return data[0] == -2; }
```
Comment (LeftofZen) [inline src/OpenLoco/Company.h:71]: what does the `-2` represent? if we can put it in a const variable with a name that would be great

File: src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#include "../CompanyManager.h"
+#include "GameCommands.h"
+
+using namespace OpenLoco::Interop;
+
+namespace OpenLoco::GameCommands
+{
+    static loco_global<CompanyId, 0x009C68EB> _updatingCompanyId;
```
Comment (LeftofZen) [inline src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp:8]: I think we want to minimise declarations of loco_global, so you can make these getters/setters in `gameState`, and then you call the getter/setter from here

Comment (GalBr) [reply src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp]: Done!

File: src/OpenLoco/Company.h (hunk 2 above)
Comment (GalBr) [reply src/OpenLoco/Company.h]: I'm not sure. `-2` was the value for `ax` when using the entityId for `cx` in the code that hooked to the original function. I guess it is used as an invalid / unused marker for that register. What do you suggest the name should be? Should it be a const variable inside the OwnerStatus class?

**Comment by GalBr** [comment]
I updated the branch with new code that answers the comment and seems to be working. Let me know what you think.

File: src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace OpenLoco::GameCommands
+{
+    // 0x004383CA
+    static uint32_t updateOwnerStatus(const uint8_t flags, OwnerStatus ownerStatus)
+    {
+        if (flags & Flags::apply)
+        {
+            auto companyId = CompanyManager::getUpdatingCompanyId();
+            auto company = CompanyManager::get(companyId);
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp:14]: As this is a networked thing we should check for nullptr on return. If it is nullptr return FAILURE

Comment (GalBr) [reply src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp:14]: Like so?

File: src/OpenLoco/Company.h [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 #include "Economy/Currency.h"
 #include "Economy/Expenditures.h"
+#include "Map/Tile.h"
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/OpenLoco/Company.h:5]: the declaration only uses a reference so it will be fine.

[the remark this answers is not in the document]
Comment (GalBr) [reply src/OpenLoco/Company.h:5]: I'm not following.
`OwnerStatus(const Map::Pos2& pos);` can't work in `Company.h` without this include. 
is there a way to make this work that I'm not aware of?

[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/OpenLoco/Company.h:5]: If Map::Pos is forward declared then it should all work.

File: src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp:14]: can you do `auto*` just to make it explicit that this is a pointer.

File: src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
+#include "../CompanyManager.h"
+#include "GameCommands.h"
+
+using namespace OpenLoco::Interop;
+
+namespace OpenLoco::GameCommands
+{
+    // 0x004383CA
+    static uint32_t updateOwnerStatus(const uint8_t flags, OwnerStatus ownerStatus)
```
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp:9]: this cant be const &

File: src/OpenLoco/Company.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (GalBr) [reply src/OpenLoco/Company.h]: Done.

File: src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp (hunk 6 above)
Comment (GalBr) [reply src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp]: Done.

File: src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp (hunk 7 above)
Comment (GalBr) [reply src/OpenLoco/GameCommands/UpdateOwnerStatus.cpp]: I assume you meant can and not can't, so I changed `OwnerStatus` to `const OwnerStatus&`

**Review by duncanspumpkin — APPROVED** [review]

File: src/OpenLoco/Company.h [hunk 8]
```diff
@@ -113,7 +146,8 @@ namespace OpenLoco
         int16_t observationY;                // 0x8BC0;
         uint16_t observationObject;          // 0x8BC2;
         uint16_t var_8BC4;
-        uint8_t pad_8BC6[0x8BCE - 0x8BC6];
+        OwnerStatus ownerStatus; // 0x8BC6
+        uint8_t pad_8BC6[0x8BCE - 0x8BCA];
```
Comment (duncanspumpkin) [inline src/OpenLoco/Company.h:150]: just spotted this today this should say `pad_8BCA`

Comment (GalBr) [reply src/OpenLoco/Company.h]: Good catch! Fixed.

**Pull request merged**